cloudpathlib/s3/s3client.py

```diff
@@ -179,25 +179,30 @@ def _list_dir(self, cloud_path: S3Path, recursive=False) -> Iterable[S3Path]:
                 for parent in PurePosixPath(o.key[len(prefix) :]).parents:
                     # if we haven't surfaced their directory already
                     if parent not in yielded_dirs and str(parent) != ".":
-                        yield self.CloudPath(f"s3://{cloud_path.bucket}/{prefix}{parent}")
+                        yield (self.CloudPath(f"s3://{cloud_path.bucket}/{prefix}{parent}"), True)
                         yielded_dirs.add(parent)
 
-                yield self.CloudPath(f"s3://{o.bucket_name}/{o.key}")
+                yield (self.CloudPath(f"s3://{o.bucket_name}/{o.key}"), False)
         else:
             # non recursive is best done with old client API rather than resource
             paginator = self.client.get_paginator("list_objects")
 
             for result in paginator.paginate(
                 Bucket=cloud_path.bucket, Prefix=prefix, Delimiter="/"
             ):
-
                 # sub directory names
                 for result_prefix in result.get("CommonPrefixes", []):
-                    yield self.CloudPath(f"s3://{cloud_path.bucket}/{result_prefix.get('Prefix')}")
+                    yield (
+                        self.CloudPath(f"s3://{cloud_path.bucket}/{result_prefix.get('Prefix')}"),
+                        True,
+                    )
 
                 # files in the directory
                 for result_key in result.get("Contents", []):
-                    yield self.CloudPath(f"s3://{cloud_path.bucket}/{result_key.get('Key')}")
+                    yield (
+                        self.CloudPath(f"s3://{cloud_path.bucket}/{result_key.get('Key')}"),
+                        False,
+                    )
```

Comment:
@pjbull: Reviewing @GeorgeSabu's #206 (thanks!) made me realize a problem here. I think we need to both check whether `result_key` is size 0 and ends with trailing `/`, similar to what was done in #190.

Reply:
Yeah, I wanted to keep fixing that issue separate, so I'm going to merge this and then rebase #206 onto it and test with some "fake" directories.